python/cufinufft/tests/test_simple.py

```diff
@@ -11,77 +11,70 @@
 
 DTYPES = [np.float32, np.float64]
 SHAPES = [(64,), (64, 64), (64, 64, 64)]
+N_TRANS = [(), (1,), (2,)]
 MS = [256, 1024, 4096]
 TOLS = [1e-2, 1e-3]
 OUTPUT_ARGS = [False, True]
 
 @pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("n_trans", N_TRANS)
+@pytest.mark.parametrize("shape", SHAPES)
 @pytest.mark.parametrize("M", MS)
 @pytest.mark.parametrize("tol", TOLS)
 @pytest.mark.parametrize("output_arg", OUTPUT_ARGS)
-def test_nufft3d1(dtype, M, tol, output_arg):
-    shape = (64, 64, 64)
-
+def test_simple_type1(dtype, shape, n_trans, M, tol, output_arg):
     real_dtype = dtype
     complex_dtype = utils._complex_dtype(dtype)
 
     dim = len(shape)
 
-    k = utils.gen_nu_pts(M, dim=dim).astype(real_dtype)
-    c = utils.gen_nonuniform_data(M).astype(complex_dtype)
+    fun = {1: cufinufft.nufft1d1,
+           2: cufinufft.nufft2d1,
+           3: cufinufft.nufft3d1}[dim]
+
+    k, c = utils.type1_problem(dtype, shape, M, n_trans=n_trans)
 
     k_gpu = gpuarray.to_gpu(k)
     c_gpu = gpuarray.to_gpu(c)
 
     if output_arg:
-        fk_gpu = gpuarray.GPUArray(shape, dtype=complex_dtype)
-        cufinufft.nufft3d1(*k_gpu, c_gpu, out=fk_gpu, eps=tol)
+        fk_gpu = gpuarray.GPUArray(n_trans + shape, dtype=complex_dtype)
+        fun(*k_gpu, c_gpu, out=fk_gpu, eps=tol)
     else:
-        fk_gpu = cufinufft.nufft3d1(*k_gpu, c_gpu, shape, eps=tol)
+        fk_gpu = fun(*k_gpu, c_gpu, shape, eps=tol)
 
     fk = fk_gpu.get()
 
-    ind = int(0.1789 * np.prod(shape))
-
-    fk_est = fk.ravel()[ind]
-    fk_target = utils.direct_type1(c, k, shape, ind)
-
-    type1_rel_err = np.abs(fk_target - fk_est) / np.abs(fk_target)
-
-    assert type1_rel_err < 10 * tol
+    utils.verify_type1(k, c, fk, tol)
 
 
 @pytest.mark.parametrize("dtype", DTYPES)
+@pytest.mark.parametrize("shape", SHAPES)
+@pytest.mark.parametrize("n_trans", N_TRANS)
 @pytest.mark.parametrize("M", MS)
 @pytest.mark.parametrize("tol", TOLS)
 @pytest.mark.parametrize("output_arg", OUTPUT_ARGS)
-def test_nufft3d2(dtype, M, tol, output_arg):
-    shape = (64, 64, 64)
-
+def test_simple_type2(dtype, shape, n_trans, M, tol, output_arg):
     real_dtype = dtype
     complex_dtype = utils._complex_dtype(dtype)
 
     dim = len(shape)
 
-    k = utils.gen_nu_pts(M, dim=dim).astype(real_dtype)
-    fk = utils.gen_uniform_data(shape).astype(complex_dtype)
+    fun = {1: cufinufft.nufft1d2,
```

Comment:
this got too abstract for me to understand!

Reply:
We're defining a dictionary `{1: nufft1d2, 2: nufft2d1, 3: nufft3d1}` mapping dimensions to function (pointers) and then indexing into it right afterwards `[dim]`. Think of it as a type of switch statement.